feat(docs): Official theme development workflow (#22177)
gatsbyjs/gatsby: docs/contributing/code-contributions.md

Check [Debugging the build process](/docs/debugging-the-build-process/) page to learn how to debug Gatsby.

## Official theme development

This section is for official theme development in Gatsby's monorepo. If you are looking
to build your own theme, see [building themes](/docs/themes/building-themes/).

Before getting started, make sure that you have
[set up your local dev environment](/contributing/setting-up-your-local-dev-environment/)
and that you're on the latest version of `gatsby-dev-cli`.

> - In the Gatsby monorepo, change directory to one of the starters using the theme you want to work on: `cd starters/gatsby-starter-blog-theme`

Reviewer: ```suggestion
- In the Gatsby monorepo find the starter in the `/starters` directory that uses the theme you want to work on
- Navigate to that directory, e.g. `cd starters/gatsby-starter-blog-theme`
```

- Install dependencies: `yarn`
- Run Gatsby Dev CLI to sync theme files: `gatsby-dev --packages gatsby-theme-blog`

> - In another tab run the starter: `yarn develop`

Reviewer: This is failing for me. On sharp though, so hard to say if it's just my local stuff.

Reviewer: Got it! Looks like it was a weird node thing.

Check [Debugging the build process](/docs/debugging-the-build-process/) page to learn how to debug Gatsby.

## Official theme development

This section is for official theme development in Gatsby's monorepo. If you are looking
to build your own theme, see [building themes](/docs/themes/building-themes/).

Before getting started, make sure that you have
[set up your local dev environment](/contributing/setting-up-your-local-dev-environment/)
and that you're on the latest version of `gatsby-dev-cli`.

> - Scaffold out a starter that uses the theme you want to work on:

Reviewer: Where should this happen? Are we talking about the starter theme in the starter repo?
Based on my testing that seems correct. So perhaps we should make clear that relative path is meant to be whatever starter that runs the theme is?

Reviewer: On closer look it seems like you're saying that a starter using the theme needs to exist, otherwise make one. Then create a site based on that starter. Is that correct? If so, we should be explicit.

Author: I realized we can simplify this by using monorepo starters, so I've updated the instructions to reflect that. Let me know if that makes things more clear!

Reviewer: Haha, that's what I did without realizing that wasn't what you meant. I feel vindicated!

- Navigate to that directory, e.g. `cd starters/gatsby-starter-blog-theme`
- Install dependencies: `yarn`

> - Run Gatsby Dev CLI to sync theme files: `gatsby-dev --packages gatsby-theme-blog`

Reviewer: ```suggestion
- Run Gatsby Dev CLI to sync theme files, referencing the appropriate theme: `gatsby-dev --packages gatsby-theme-blog`
```

- In another tab run the starter with `yarn develop`

> - Edit the theme files in your local gatsby monorepo (where you set your `gatsby-dev` path). You'll see changes automatically copied over and update in your starter.

Reviewer: ```suggestion
- Edit the theme files in your local Gatsby monorepo (where you set your `gatsby-dev` path). You'll see changes automatically copied over and update in your starter.
```
